diff --git a/lib/parse.js b/lib/parse.js
index 3a6988e629f52aee6609dc5ec290a5e9ff47cecb..cbaa90e1f4336f57e1092512e2f40551e300b960 100644
--- a/lib/parse.js
+++ b/lib/parse.js
@@ -143,6 +143,7 @@ const parse = (input, options = {}) => {
      */
 
     if (value === CHAR_LEFT_PARENTHESES) {
+      if (stack.length >= 64) return { type: 'root', input, nodes: [{ type: 'text', value: input }] };
       block = push({ type: 'paren', nodes: [] });
       stack.push(block);
       push({ type: 'text', value });
@@ -195,6 +196,7 @@ const parse = (input, options = {}) => {
      */
 
     if (value === CHAR_LEFT_CURLY_BRACE) {
+      if (stack.length >= 64) return { type: 'root', input, nodes: [{ type: 'text', value: input }] };
       depth++;
 
       const dollar = prev.value && prev.value.slice(-1) === '$' || block.dollar === true;
